### PR TITLE
Fix: Exit with error code for testing Dummy Agent in CI

### PR DESCRIPTION
**Pull request by SmartManoj**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**
Fixes https://github.com/OpenDevin/OpenDevin/issues/2618

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

Checked for exit code in a single step.

Unlike in #2834, The "Run E2E test with dummy agent" workflow should fail.

### CODE REVIEW

**Review by tobitege — APPROVED** [review]

**Review by enyst — CHANGES_REQUESTED** [review]
I'm not sure what the effects of these changes are. Can you please tell, has it been tested, for example:
- with UI, agent gets some error state; before this PR and after, can the user continue and have the agent start a new task?
- without UI, a task gets an error and finishes execution; the user restarts, loading the previous session; before this PR and after: does the new task work and end normally when the user is sending "/exit"?

**Comment by SmartManoj** [comment]
> * with UI, agent gets some error state; before this PR and after, can the user continue and have the agent start a new task?

`main.py` is not used for UI. right?

> * without UI, a task gets an error and finishes execution; the user restarts, loading the previous session; before this PR and after: does the new task work and end normally when the user is sending "/exit"?

Why not? It doesn't change any state of the session.

**Review by tobitege — APPROVED** [review]
LGTM

**Comment by tobitege** [comment]
@enyst the changes make sense to me, improved error handling

@SmartManoj just need to fix that small conflict

**Comment by enyst** [comment]
@tobitege @SmartManoj When restoring the session in cli, it will exit with an error all the time, if it was one time. Because `last_error` is never cleared, I think.

I'll double-check and I can submit a PR to fix that behavior. I'll need to do some testing, it's used in evals. Status of a single task as run in evals was the only purpose of this `last_error` attribute, it was a bit of a hack and it doesn't - yet - behave the way we seem to assume it does.

**Review by enyst — APPROVED** [review]
That was fixed, now using `last_error` works with sessions as expected.

**Comment by tobitege** [comment]
Issue fixed by #3146 , closing this PR.

**Comment by SmartManoj** [comment]
> Issue fixed by #3146

Nope, that is another bug.

**Comment by tobitege** [comment]
Please revert the return type hint for `run_agent_controller` back to `State | None`.

Then deal with the return value accordingly:
```py
    final_state = asyncio.run(
        run_agent_controller(
            agent=agent,
            task_str=task_str,
            max_iterations=args.max_iterations,
            max_budget_per_task=args.max_budget_per_task,
        )
    )
    if final_state and hasattr(final_state, 'last_error') and final_state.last_error:
        logger.error(f"Agent encountered an error: {final_state.last_error}")
        sys.exit(1)
```

**Review by tobitege — DISMISSED** [review]
See my last comment for changes.

**Comment by SmartManoj** [comment]
> Please revert the return type hint for `run_agent_controller` back to `State | None`.

When it returns `None`?

**Comment by tobitege** [comment]
> > Please revert the return type hint for `run_agent_controller` back to `State | None`.
> 
> When it returns `None`?

Can you first ensure that all possible execution paths in the function return a State object, and that there are **no** cases where it could return None and never will?

**Comment by SmartManoj** [comment]
> > > Please revert the return type hint for `run_agent_controller` back to `State | None`.
> > 
> > 
> > When it returns `None`?
> 
> Can you first ensure that all possible execution paths in the function return a State object, and that there are **no** cases where it could return None and never will?

After ensuring only, I removed that type, and `mypy` also accepted it.

File: opendevin/core/main.py [hunk 1]
```diff
@@ -184,11 +184,13 @@ async def on_event(event: Event):
...
+    final_state = asyncio.run(
         run_agent_controller(
             agent=agent,
             task_str=task_str,
             max_iterations=args.max_iterations,
             max_budget_per_task=args.max_budget_per_task,
         )
     )
+    if final_state.last_error:
+        sys.exit(1)
```
Comment (li-boxuan) [inline opendevin/core/main.py:196]: Please don't do this. We should only use `sys.exit(1)` in production code (tests excluded) as last-resort.

**Review by li-boxuan — CHANGES_REQUESTED** [review]
A proper way to fix this is to write a python script that runs main function and check final state.

**Comment by SmartManoj** [comment]
> A proper way to fix this is to write a python script that runs main function and check final state.

By duplicating the `main` block?

File: opendevin/core/main.py (hunk 1 above)
Comment (SmartManoj) [reply opendevin/core/main.py:196]: side effects of this? It just sets the current exit code of the script.

**Comment by li-boxuan** [comment]
> > A proper way to fix this is to write a python script that runs main function and check final state.
> 
> By duplicating the `main` block?

Take a look at `test_agent.py`. You could call main function.

File: tests/integration/test_agent.py [hunk 2]
```diff
@@ -282,3 +281,15 @@ def test_browse_internet(http_server, current_test_name: str):
         assert 'OpenDevin is all you need!' in last_observation.content
     elif isinstance(last_observation, AgentDelegateObservation):
         assert 'OpenDevin is all you need!' in last_observation.outputs['content']
+
+
+def test_dummy_agent(current_test_name: str):
```
Comment (SmartManoj) [inline tests/integration/test_agent.py:286]: https://github.com/OpenDevin/OpenDevin/pull/1316#discussion_r1581732909 @li-boxuan Why do you think it's a bad idea to add this as a pytest?

Comment (li-boxuan) [reply tests/integration/test_agent.py:287]: I don't remember. Things might have changed and maybe it's a good time now.

Comment (li-boxuan) [reply tests/integration/test_agent.py:287]: You need to do ```@pytest.mark.skipif(os.environ.get('DEFAULT_AGENT') != 'DummyAgent')``` for this test

and also skip DummyAgent for every other test.

File: tests/integration/test_agent.py [hunk 3]
```diff
@@ -282,3 +282,15 @@ def test_browse_internet(http_server, current_test_name: str):
...
```
Comment (li-boxuan) [inline tests/integration/test_agent.py:287]: You need to do ```@pytest.mark.skipif(os.environ.get('DEFAULT_AGENT') != 'DummyAgent')``` for this test

and also skip DummyAgent for every other test.

Comment (SmartManoj) [reply tests/integration/test_agent.py:287]: how about separate `test_dummy_agent.py`?

**Comment by rbren** [comment]
We're doing some cleanup of PRs, and are going to close this one for now.

File: opendevin/core/main.py (hunk 1 above)
Comment (BradKML) [reply opendevin/core/main.py]: @li-boxuan what is the alternative/standard in ending an agent?

**Pull request closed without merging**